---
title: "Transcriptomic analysis of yeast in response to selective nutrient limitation"
author: "Gabe Mednick"
date: "2021-11-11"
output:
  html_document:
    df_print: paged
categories: []
tags: []
subtitle: ''
summary: This post demonstrates how the Tidyvere can enlighten any omics analysis
authors: []
lastmod: ''
featured: no
image:
  caption: ''
  focal_point: ''
  preview_only: no
projects: []
slug: yeast_expression
---

<script src="{{< blogdown/postref >}}index_files/header-attrs/header-attrs.js"></script>
<script src="{{< blogdown/postref >}}index_files/htmlwidgets/htmlwidgets.js"></script>
<link href="{{< blogdown/postref >}}index_files/datatables-css/datatables-crosstalk.css" rel="stylesheet" />
<script src="{{< blogdown/postref >}}index_files/datatables-binding/datatables.js"></script>
<script src="{{< blogdown/postref >}}index_files/jquery/jquery-3.6.0.min.js"></script>
<link href="{{< blogdown/postref >}}index_files/dt-core/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="{{< blogdown/postref >}}index_files/dt-core/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="{{< blogdown/postref >}}index_files/dt-core/js/jquery.dataTables.min.js"></script>
<link href="{{< blogdown/postref >}}index_files/crosstalk/css/crosstalk.min.css" rel="stylesheet" />
<script src="{{< blogdown/postref >}}index_files/crosstalk/js/crosstalk.min.js"></script>


<div id="this-post-is-still-under-construction" class="section level3">
<h3>This post is still under construction</h3>
</div>
<div id="yeast-microarray-data" class="section level2">
<h2>Yeast microarray data</h2>
<p>In biology, the term ‘omics’ is used to refer to studies that involve a meta view of the behavior of an entire class of molecules. Omics can refer to a range of disciplines, including genomics, proteomics, metabolomics, metagenomics and transcriptomics. This post will focus on transcriptomics, the field of study that looks at RNA expression and transcriptional regulation in response to specific changes. We will be using gene expression data from <a href="https://www.molbiolcell.org/doi/full/10.1091/mbc.e07-08-0779">Brauer et al, 2008</a>. The title of their paper provides a powerful summary of the experiment and what they learned from the data: ‘Coordination of Growth Rate, Cell Cycle, Stress Response, and Metabolic Activity in Yeast’.</p>
<p>In my last two posts, I have tried to emphasize the use of Bioconductor packages and the Tidyverse in a seamless analysis workflow. Although we may ultimately be working with specialized bioinformatics packages such as <code>Deseq2</code>, <code>EdgeR</code> or <code>Lima</code> for analysis of transcriptomic data, David Robinson notes that ‘Any challenge in plotting, filtering, or merging your data will get directly in the way of answering scientific questions’. This is why the <code>Tidyverse</code> is essential to any analysis workflow.</p>
<p>As noted, the data is from yeast and the gene expression levels were measured under conditions of selective nutrient starvation. Growth rates were also carefully controlled by using a chemostat to regulate the flow of fresh growth media. The expression data was collected using a microarray. Microarrays measure expression levels as fluorescence intensities that are proportional to the number of sequences that bind to the complementary probe sequences of the array. RNA-seq experiments are considered more accurate since all RNA transcripts are counted directly.</p>
<p>Before we dive into the deeps of this analysis, I want to give credit and thanks to David Robinson. David is one of the most insightful and inspiring data scientists on the planet. He has a few courses on DataCamp and regularly <a href="https://www.youtube.com/channel/UCeiiqmVK07qhY-wvg3IZiZQ">livestreams</a> data analysis with R, the Tidyverse and Tidymodels. This post will closely follow a <a href="http://varianceexplained.org/r/tidy-genomics-biobroom/">series of posts</a> by David that focus on working with genomic data.</p>
<p>Let’s start by taking a look at the first few rows of data</p>
<pre class="r"><code># yeast_data &lt;- read_delim(&quot;http://varianceexplained.org/files/Brauer2008_DataSet1.tds&quot;)
# write_rds(yeast_data, &#39;yeast_data&#39;)
yeast_df &lt;- read_rds(&quot;yeast_data&quot;)

# yeast_df %&gt;% 
#   glimpse()

yeast_df %&gt;% 
  head(5) %&gt;% 
  gt() %&gt;% 
  tab_style(
    style = list(
      cell_fill(color = &quot;#E0FFFF&quot;),
      cell_text(weight = &quot;bold&quot;)
      ),
    locations = cells_body())</code></pre>
<div id="kcuhefrjln" style="overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>html {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Helvetica Neue', 'Fira Sans', 'Droid Sans', Arial, sans-serif;
}

#kcuhefrjln .gt_table {
  display: table;
  border-collapse: collapse;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#kcuhefrjln .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#kcuhefrjln .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#kcuhefrjln .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 0;
  padding-bottom: 6px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#kcuhefrjln .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#kcuhefrjln .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#kcuhefrjln .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#kcuhefrjln .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#kcuhefrjln .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#kcuhefrjln .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#kcuhefrjln .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#kcuhefrjln .gt_group_heading {
  padding: 8px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
}

#kcuhefrjln .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#kcuhefrjln .gt_from_md > :first-child {
  margin-top: 0;
}

#kcuhefrjln .gt_from_md > :last-child {
  margin-bottom: 0;
}

#kcuhefrjln .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#kcuhefrjln .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 12px;
}

#kcuhefrjln .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#kcuhefrjln .gt_first_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
}

#kcuhefrjln .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#kcuhefrjln .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#kcuhefrjln .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#kcuhefrjln .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#kcuhefrjln .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#kcuhefrjln .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding: 4px;
}

#kcuhefrjln .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#kcuhefrjln .gt_sourcenote {
  font-size: 90%;
  padding: 4px;
}

#kcuhefrjln .gt_left {
  text-align: left;
}

#kcuhefrjln .gt_center {
  text-align: center;
}

#kcuhefrjln .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#kcuhefrjln .gt_font_normal {
  font-weight: normal;
}

#kcuhefrjln .gt_font_bold {
  font-weight: bold;
}

#kcuhefrjln .gt_font_italic {
  font-style: italic;
}

#kcuhefrjln .gt_super {
  font-size: 65%;
}

#kcuhefrjln .gt_footnote_marks {
  font-style: italic;
  font-weight: normal;
  font-size: 65%;
}
</style>
<table class="gt_table">
  
  <thead class="gt_col_headings">
    <tr>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">GID</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">YORF</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">NAME</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">GWEIGHT</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">G0.05</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">G0.1</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">G0.15</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">G0.2</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">G0.25</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">G0.3</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">N0.05</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">N0.1</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">N0.15</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">N0.2</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">N0.25</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">N0.3</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">P0.05</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">P0.1</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">P0.15</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">P0.2</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">P0.25</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">P0.3</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">S0.05</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">S0.1</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">S0.15</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">S0.2</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">S0.25</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">S0.3</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">L0.05</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">L0.1</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">L0.15</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">L0.2</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">L0.25</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">L0.3</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">U0.05</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">U0.1</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">U0.15</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">U0.2</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">U0.25</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">U0.3</th>
    </tr>
  </thead>
  <tbody class="gt_table_body">
    <tr><td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">GENE1331X</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">A_06_P5820</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">SFB2       || ER to Golgi transport || molecular function unknown || YNL049C || 1082129</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">1</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.24</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.13</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.21</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.15</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.05</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.05</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.20</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.24</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.20</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.42</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.14</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.09</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.26</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.20</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.22</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.31</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.04</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.34</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.51</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.12</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.09</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.09</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.20</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.08</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.18</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.18</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.13</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.20</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.17</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.11</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.06</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.26</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.05</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.28</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.19</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.09</td></tr>
    <tr><td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">GENE4924X</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">A_06_P5866</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">          || biological process unknown || molecular function unknown || YNL095C || 1086222</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">1</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.28</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.13</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.40</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.48</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.11</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.17</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.31</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.00</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.63</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.44</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.26</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.21</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.09</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.04</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.10</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.15</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.20</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.63</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.53</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.15</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.01</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.12</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.15</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.32</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.16</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.09</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.02</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.04</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.03</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.01</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-1.02</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.91</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.59</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.61</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.17</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.18</td></tr>
    <tr><td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">GENE4690X</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">A_06_P1834</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">QRI7       || proteolysis and peptidolysis || metalloendopeptidase activity || YDL104C || 1085955</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">1</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.02</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.27</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.27</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.02</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.24</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.25</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.23</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.06</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.66</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.40</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.46</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.43</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.18</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.22</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.33</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.34</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.13</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.44</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">1.29</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.32</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.47</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.50</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.42</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.33</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.30</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.02</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.07</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.05</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.13</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.04</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.91</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.94</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.42</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.36</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.49</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.47</td></tr>
    <tr><td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">GENE1177X</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">A_06_P4928</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">CFT2       || mRNA polyadenylylation* || RNA binding || YLR115W || 1081958</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">1</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.33</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.41</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.24</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.03</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.03</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.00</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.20</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.25</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.49</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.49</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.43</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.26</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.05</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.04</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.03</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.04</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.08</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.21</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.41</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.43</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.21</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.33</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.05</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.24</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.27</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.28</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.05</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.02</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.00</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.08</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.53</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.51</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.26</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.05</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.14</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.01</td></tr>
    <tr><td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">GENE511X</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">A_06_P5620</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">SSO2       || vesicle fusion* || t-SNARE activity || YMR183C || 1081214</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">1</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.05</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.02</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.40</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.34</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.13</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.14</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.35</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.09</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.08</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.58</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.14</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.12</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.16</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.18</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.21</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.08</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.23</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.29</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.70</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.05</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.10</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.07</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.10</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.32</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.59</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.13</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.00</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.11</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.04</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.01</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.45</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.09</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.13</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.02</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.09</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.03</td></tr>
  </tbody>
  
  
</table>
</div>
</div>
<div id="cleaning-the-data" class="section level2">
<h2>Cleaning the data</h2>
<p>Now that we’ve seen the data, let’s talk about how we can make the information more useful by tidying it. A Tidy data frame should have one variable per column and one observation per row. In the Brauer data, the <code>NAME</code> column has multiple pieces of information separated by ‘||’. Similarly, the numeric column names are made up of the nutrient that is being limited in growth media and the growth rate. The respective values are the relative expression levels for each gene under the given conditions. Enough talk, let’s code.</p>
<pre class="r"><code># I&#39;m going to pivot the table longer by combining all nutrient conditions into one column and expression values into another. Then we can separate and the information in the above mentioned columns in only a couple of steps using separate()
yeast_df_tidy &lt;- yeast_df %&gt;% 
  pivot_longer(G0.05:U0.3, names_to = &#39;nutrient_growth&#39;, values_to = &#39;expression_levels&#39;) %&gt;% 
  separate(nutrient_growth, c(&quot;nutrient&quot;, &quot;growth_rate&quot;), sep = 1, convert = TRUE) %&gt;% 
  separate(NAME, c(&#39;gene_name&#39;, &#39;molec_process&#39;, &#39;activity&#39;, &#39;sys_id&#39;, &#39;id&#39;), sep = &quot;\\|\\|&quot;) %&gt;% 
  mutate(across(everything(), str_squish),
         growth_rate = as.numeric(growth_rate),
         expression_levels = as.numeric(expression_levels),
         activity = str_remove(molec_process, &quot;[*]&quot;),
         molec_process = str_remove(activity, &quot;[*]&quot;),
         molec_process = str_remove(activity, &quot;activity&quot;)) %&gt;% 
  janitor::clean_names() # clean column names
# growth_rate and expression_levels are characters--&gt;let&#39;s convert them to numeric </code></pre>
<p>Let’s look at the data again after cleaning it!</p>
<pre class="r"><code>yeast_df_tidy %&gt;% 
  head(5) %&gt;% 
  gt() %&gt;% 
  tab_style(
    style = list(
      cell_fill(color = &quot;#E0FFFF&quot;),
      cell_text(weight = &quot;bold&quot;)
      ),
    locations = cells_body())</code></pre>
<div id="zkeivpcprd" style="overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>html {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Helvetica Neue', 'Fira Sans', 'Droid Sans', Arial, sans-serif;
}

#zkeivpcprd .gt_table {
  display: table;
  border-collapse: collapse;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#zkeivpcprd .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#zkeivpcprd .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#zkeivpcprd .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 0;
  padding-bottom: 6px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#zkeivpcprd .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#zkeivpcprd .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#zkeivpcprd .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#zkeivpcprd .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#zkeivpcprd .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#zkeivpcprd .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#zkeivpcprd .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#zkeivpcprd .gt_group_heading {
  padding: 8px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
}

#zkeivpcprd .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#zkeivpcprd .gt_from_md > :first-child {
  margin-top: 0;
}

#zkeivpcprd .gt_from_md > :last-child {
  margin-bottom: 0;
}

#zkeivpcprd .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#zkeivpcprd .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 12px;
}

#zkeivpcprd .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#zkeivpcprd .gt_first_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
}

#zkeivpcprd .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#zkeivpcprd .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#zkeivpcprd .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#zkeivpcprd .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#zkeivpcprd .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#zkeivpcprd .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding: 4px;
}

#zkeivpcprd .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#zkeivpcprd .gt_sourcenote {
  font-size: 90%;
  padding: 4px;
}

#zkeivpcprd .gt_left {
  text-align: left;
}

#zkeivpcprd .gt_center {
  text-align: center;
}

#zkeivpcprd .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#zkeivpcprd .gt_font_normal {
  font-weight: normal;
}

#zkeivpcprd .gt_font_bold {
  font-weight: bold;
}

#zkeivpcprd .gt_font_italic {
  font-style: italic;
}

#zkeivpcprd .gt_super {
  font-size: 65%;
}

#zkeivpcprd .gt_footnote_marks {
  font-style: italic;
  font-weight: normal;
  font-size: 65%;
}
</style>
<table class="gt_table">
  
  <thead class="gt_col_headings">
    <tr>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">gid</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">yorf</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">gene_name</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">molec_process</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">activity</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">sys_id</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">id</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">gweight</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">nutrient</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">growth_rate</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">expression_levels</th>
    </tr>
  </thead>
  <tbody class="gt_table_body">
    <tr><td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">GENE1331X</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">A_06_P5820</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">SFB2</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">ER to Golgi transport</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">ER to Golgi transport</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">YNL049C</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">1082129</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">1</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">G</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.05</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.24</td></tr>
    <tr><td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">GENE1331X</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">A_06_P5820</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">SFB2</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">ER to Golgi transport</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">ER to Golgi transport</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">YNL049C</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">1082129</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">1</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">G</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.10</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.13</td></tr>
    <tr><td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">GENE1331X</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">A_06_P5820</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">SFB2</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">ER to Golgi transport</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">ER to Golgi transport</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">YNL049C</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">1082129</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">1</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">G</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.15</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.21</td></tr>
    <tr><td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">GENE1331X</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">A_06_P5820</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">SFB2</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">ER to Golgi transport</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">ER to Golgi transport</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">YNL049C</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">1082129</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">1</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">G</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.20</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.15</td></tr>
    <tr><td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">GENE1331X</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">A_06_P5820</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">SFB2</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">ER to Golgi transport</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">ER to Golgi transport</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">YNL049C</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">1082129</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">1</td>
<td class="gt_row gt_left" style="background-color: #E0FFFF; font-weight: bold;">G</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">0.25</td>
<td class="gt_row gt_right" style="background-color: #E0FFFF; font-weight: bold;">-0.05</td></tr>
  </tbody>
  
  
</table>
</div>
<p>Cool, let’s explore some of the benefits of Tidying our data? For starters, we can explore the data with the data visualization package, <code>GGplot2</code>. When we cleaned the data, we separated important annotation information about each gene. Let’s use this information to count each category and see what the most common molecular processes in .</p>
<pre class="r"><code># plotting most common gene product associated molecular processes
yeast_df_tidy %&gt;% 
  select(gene_name, activity, molec_process, nutrient, growth_rate, expression_levels) %&gt;% 
  drop_na(molec_process) %&gt;%
  group_by(gene_name, molec_process) %&gt;% 
  count(sort = T) %&gt;%
  ungroup() %&gt;%
  slice(n = 4:20) %&gt;% 
  ggplot(aes(n, fct_reorder(molec_process, n), fill = molec_process)) +
  geom_col() +
  theme(legend.position = &#39;none&#39;) +
  labs(title = &#39;Most common molecular processes&#39;,
       subtitle = &quot;Gene count in yeast&quot;,
       y = NULL,
       x = NULL) +
  theme(plot.title = element_text(color = &quot;blue&quot;, size = 14, face = &quot;bold&quot;, hjust = -1))</code></pre>
<p><img src="{{< blogdown/postref >}}index_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<p>Let’s also look at the annotation for gene product activity.</p>
<pre class="r"><code># plotting most common gene product function
yeast_df_tidy %&gt;% 
  select(gene_name, activity, molec_process, nutrient, growth_rate, expression_levels) %&gt;% 
  drop_na(activity) %&gt;%
  group_by(gene_name, activity) %&gt;% 
  count(sort = T) %&gt;%
  ungroup() %&gt;%
  slice(n = 4:20) %&gt;% 
  ggplot(aes(n,fct_reorder(activity, n), fill = activity)) +
  geom_col() +
  theme(legend.position = &#39;none&#39;) +
  labs(title = &quot;Most common gene product activity&quot;,
       subtitle = &quot;Gene count in yeast&quot;,
       y = NULL,
       x = NULL) +
  theme(plot.title = element_text(color = &quot;blue&quot;, size = 14, face = &quot;bold&quot;, hjust = -1))</code></pre>
<p><img src="{{< blogdown/postref >}}index_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<p>Visualizing relationships in our data can help us check for inconsistencies and deepen our intuition about the relationships between variables. For instance, let’s consider how expression levels depend on growth rate for each of the nutrient limiting categories. Genes that show an increase or decrease in expression may be directly affected by the specific nutrient limitation. Let’s start off by looking at a couple of individual genes under each nutrient limiting condition and then expand our visual exploration to look for patterns based on activity or molecular process.</p>
<pre class="r"><code>#Let&#39;s replace the nutrient symbols with names
nutrients &lt;- c(G = &#39;Glucose&#39;, L = &#39;Leucine&#39;, N = &#39;Nitrogen&#39;, P  = &#39;Phosphate&#39;, S = 
&#39;Sulfate&#39;, N = &#39;Ammonia&#39;, U = &#39;Uracil&#39;)

yeast_df_tidy &lt;- yeast_df_tidy %&gt;% 
  mutate(nutrient = recode(
    nutrient, !!!nutrients
  ))</code></pre>
<pre class="r"><code>yeast_df_tidy %&gt;% 
  select(gene_name, activity, molec_process, nutrient, growth_rate, expression_levels) %&gt;% 
  filter(gene_name == &quot;SUL1&quot;) %&gt;% 
  ggplot(aes(growth_rate, expression_levels, color = nutrient)) +
  geom_line() +
  labs(title = &quot;Expression levels vs growth rate for SUL1&quot;,
       subtitle = &quot;SUL1 is involved in sulfate transport&quot;,
       x = &quot;Growth rate&quot;,
       y = &quot;Expression&quot;)</code></pre>
<p><img src="{{< blogdown/postref >}}index_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<p>Without more information about SUL1, it’s hard to know what we should expect. I would expect it’s activity to increase with growth rate under sulfate limiting conditions (more sulfate present in the media). Instead, it appears to increase wit increasing uracil and phosphate.</p>
<p>This time, let’s try a gene whose protein product is directly involved in the synthesis of a limiting nutrient. For this, we will use LEU1 which is involved in the synthesis of leucine.</p>
<pre class="r"><code>yeast_df_tidy %&gt;% 
  select(gene_name, activity, molec_process, nutrient, growth_rate, expression_levels) %&gt;% 
  filter(gene_name == &quot;LEU1&quot;) %&gt;% 
  ggplot(aes(growth_rate, expression_levels, color = nutrient)) +
  geom_line() +
  labs(title = &quot;Expression levels vs growth rate for LEU1&quot;,
       x = &quot;Growth rate&quot;,
       y = &quot;Expression&quot;)</code></pre>
<p><img src="{{< blogdown/postref >}}index_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
<p>Now, we can see a very clear trend: As growth rate increasing (increasing availability of leucine in the growth media) the in-house machinery for leucine biosynthesis is turned down.</p>
<p>Let’s use <code>filter()</code> and <code>facet_wrap()</code> to search for gene products involved in different categories to get a better sense of our data.</p>
<pre class="r"><code>yeast_df_tidy %&gt;% 
  select(gid, gene_name, activity, molec_process, nutrient, growth_rate, expression_levels) %&gt;%
  filter(grepl(&quot;leucine biosynthesis&quot;, activity),
         gene_name != &quot;&quot;) %&gt;%
  group_by(gene_name, gid, nutrient) %&gt;% 
  ungroup() %&gt;% 
  distinct(gene_name, .keep_all = T) %&gt;% 
  select(gene_name, activity, molec_process, nutrient) %&gt;% 
  DT::datatable() </code></pre>
<div id="htmlwidget-1" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-1">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4"],["LEU9","LEU1","LEU2","LEU4"],["leucine biosynthesis","leucine biosynthesis","leucine biosynthesis","leucine biosynthesis"],["leucine biosynthesis","leucine biosynthesis","leucine biosynthesis","leucine biosynthesis"],["Glucose","Glucose","Glucose","Glucose"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>gene_name<\/th>\n      <th>activity<\/th>\n      <th>molec_process<\/th>\n      <th>nutrient<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"order":[],"autoWidth":false,"orderClasses":false,"columnDefs":[{"orderable":false,"targets":0}]}},"evals":[],"jsHooks":[]}</script>
<pre class="r"><code>yeast_df_tidy %&gt;% 
  select(gid, gene_name, activity, molec_process, nutrient, growth_rate, expression_levels) %&gt;%
  filter(grepl(&quot;leucine biosynthesis&quot;, activity),
         gene_name != &quot;&quot;) %&gt;%
  group_by(gene_name, gid, nutrient) %&gt;% 
  #group_slice(1:35) %&gt;% 
  ggplot(aes(growth_rate, expression_levels, color = nutrient, group = nutrient)) +
  geom_line() +
  facet_wrap(~gene_name, scales = &#39;free&#39;) +
  labs(title = &quot;Expression levels vs growth rate under selective nutrient limitation&quot;,
       subtitle = &quot;filtering for genes with &#39;leucine biosynthesis&#39; activity&quot;,
       x = &quot;Growth rate&quot;,
       y = &quot;Expression&quot;) +                                                        
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank())</code></pre>
<p><img src="{{< blogdown/postref >}}index_files/figure-html/unnamed-chunk-9-2.png" width="672" /></p>
<pre class="r"><code># let&#39;s start by looking at the data
yeast_df_tidy %&gt;% 
  select(gid, gene_name, activity, molec_process, nutrient, growth_rate, expression_levels) %&gt;%
  filter(grepl(&quot;glucose metabolism&quot;, activity),
         gene_name != &quot;&quot;) %&gt;%
  count(gene_name, molec_process, activity) %&gt;% 
  select(-n) %&gt;% 
  DT::datatable() </code></pre>
<div id="htmlwidget-2" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-2">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6"],["FBP26","GLK1","GND2","HXK1","RGT1","TOS3"],["glucose metabolism","glucose metabolism","glucose metabolism","glucose metabolism","glucose metabolism","glucose metabolism"],["glucose metabolism","glucose metabolism","glucose metabolism","glucose metabolism","glucose metabolism","glucose metabolism"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>gene_name<\/th>\n      <th>molec_process<\/th>\n      <th>activity<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"order":[],"autoWidth":false,"orderClasses":false,"columnDefs":[{"orderable":false,"targets":0}]}},"evals":[],"jsHooks":[]}</script>
<pre class="r"><code># Let&#39;s visualize genes with glucose metabolism activity
yeast_df_tidy %&gt;% 
  select(gid, gene_name, activity, molec_process, nutrient, growth_rate, expression_levels) %&gt;%
  filter(grepl(&quot;glucose metabolism&quot;, activity),
         gene_name != &quot;&quot;) %&gt;%
  add_count(gene_name) %&gt;%
  group_by(gene_name, gid) %&gt;%
  ggplot(aes(growth_rate, expression_levels, color = nutrient, group = nutrient)) +
  geom_line() +
  facet_wrap(~gene_name, scales = &#39;free&#39;) +
  labs(title = &quot;Expression levels vs growth rate under selective nutrient limitation&quot;,
       subtitle = &quot;Genes associated with glucose metabolism&quot;,
       x = &quot;Growth rate&quot;,
       y = &quot;Expression&quot;) +                                                        
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank())</code></pre>
<p><img src="{{< blogdown/postref >}}index_files/figure-html/unnamed-chunk-10-2.png" width="672" /></p>
<pre class="r"><code># group_slice function
group_slice &lt;- function(dt, v) {
  stopifnot(&#39;data.frame&#39; %in% class(dt))
  stopifnot(v &gt; 0, (v %% 1) == 0)
  if(n_groups(dt) &gt;= 2) #warn(&#39;Only 1 group to slice&#39;)
  #get the grouping variables
  groups=
    unique(dt %&gt;% select()) %&gt;%
    group_by() %&gt;%
    slice(v)
 #subset our original data
  out = inner_join(dt, groups)
  out
}

#Table for a subset of genes related to glucose metabolism
yeast_df_tidy %&gt;% 
  select(gid, gene_name, activity, molec_process, nutrient, growth_rate, expression_levels) %&gt;%
  filter(grepl(&quot;nitrogen&quot;, activity),
         gene_name != &quot;&quot;) %&gt;%
  group_by(gene_name, gid, nutrient) %&gt;% 
  #group_slice(1:35) %&gt;%
  ungroup() %&gt;% 
  distinct(gene_name, .keep_all = T) %&gt;% 
  select(gene_name, activity, molec_process, nutrient) %&gt;% 
  DT::datatable() </code></pre>
<div id="htmlwidget-3" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-3">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8"],["URE2","GLN1","DCG1","UGA1","GDH2","MKS1","NPR1","AAT2"],["regulation of nitrogen utilization","nitrogen compound metabolism","nitrogen compound metabolism","nitrogen utilization","nitrogen compound metabolism","regulation of nitrogen utilization","regulation of nitrogen utilization","nitrogen compound metabolism"],["regulation of nitrogen utilization","nitrogen compound metabolism","nitrogen compound metabolism","nitrogen utilization","nitrogen compound metabolism","regulation of nitrogen utilization","regulation of nitrogen utilization","nitrogen compound metabolism"],["Glucose","Glucose","Glucose","Glucose","Glucose","Glucose","Glucose","Glucose"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>gene_name<\/th>\n      <th>activity<\/th>\n      <th>molec_process<\/th>\n      <th>nutrient<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"order":[],"autoWidth":false,"orderClasses":false,"columnDefs":[{"orderable":false,"targets":0}]}},"evals":[],"jsHooks":[]}</script>
<pre class="r"><code>yeast_df_tidy %&gt;% 
  select(gid, gene_name, activity, molec_process, nutrient, growth_rate, expression_levels) %&gt;%
  #mutate(across(everything(), str_squish)) %&gt;%
  filter(grepl(&quot;nitrogen&quot;, activity),
         gene_name != &quot;&quot;) %&gt;%
  group_by(gene_name, gid, nutrient) %&gt;% 
  #group_slice(1:35) %&gt;% 
  ggplot(aes(growth_rate, expression_levels, color = nutrient, group = nutrient)) +
  geom_line() +
  facet_wrap(~gene_name, scales = &#39;free&#39;) +
  labs(title = &quot;Expression levels vs growth rate under selective nutrient limitation&quot;,
       subtitle = &quot;filtering for genes with &#39;Nitrogen biosynthesis&#39; activity&quot;,
       x = &quot;Growth rate&quot;,
       y = &quot;Expression&quot;) +                                                        
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank())</code></pre>
<p><img src="{{< blogdown/postref >}}index_files/figure-html/unnamed-chunk-11-2.png" width="672" /></p>
<p>In this next plot, we are going to ‘zoom’ in on one gene and visualize growth rate vs expression as a scatter plot. This time, we will consider each nutrient condition separately using <code>facet_wrap()</code> and fit a linear model to the data points with <code>geom_smooth(method = 'lm')</code>.</p>
<pre class="r"><code># Use a named character vector for unquote splicing with !!!
# level_key &lt;- c(a = &quot;apple&quot;, b = &quot;banana&quot;, c = &quot;carrot&quot;)
# recode(char_vec, !!!level_key)

yeast_df_tidy %&gt;% 
  mutate(nutrient = recode(nutrient, !!!nutrients)) %&gt;% 
  filter(gene_name == &quot;FBP26&quot;) %&gt;%
  ggplot(aes(growth_rate, expression_levels, color = nutrient)) +
  geom_point(size = 2) +
  facet_wrap(~ nutrient) +
  geom_smooth(method = &#39;lm&#39;, size = .5) +
  labs(title = &#39;FBP26 expression vs growth rate under selective nutrient starvation&#39;,
       x = &quot;Growth rate&quot;,
       y = &quot;Expression levels&quot;)</code></pre>
<p><img src="{{< blogdown/postref >}}index_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<p>FBP26 exhibits a trend of decreasing expression with increasing growth rate. We see that a linear model fits this trend quite well. In th next section we are going to generate linear models for each gene and nutrient combination. Then we will look at how we can visualize significant genes based on whether they were up or down regulated in relation to growth rate.</p>
</div>
<div id="quantifying-expression-levels-with-linear-regression" class="section level2">
<h2>Quantifying expression levels with linear regression</h2>
<p>We are going to use base R’s <code>lm()</code> function to estimate the expression for each gene-nutrient combination. This can be done by grouping and nesting the data, then using the <code>purrr</code> package to iterate over the samples and return a linear model for each. Finally, we can use the <code>broom</code> package to return a tibble and <code>ggplot2</code> to look at the results.</p>
<pre class="r"><code>library(tidymodels)
library(broom)
library(patchwork)
library(glue)

#David&#39;s example of running a linear model for each gene
# linear_models &lt;- yeast_df_tidy %&gt;%
#   drop_na(expression_levels) %&gt;% 
#   group_by(gene_name, gid, nutrient) %&gt;%
#   do(tidy(lm(expression_levels ~ growth_rate, .)))

lm_express_mod &lt;- yeast_df_tidy %&gt;% 
  drop_na() %&gt;%
  select(gene_name, gid, nutrient, expression_levels, growth_rate) %&gt;% 
  group_by(gene_name, gid, nutrient) %&gt;% 
  nest(-gene_name, -gid, -nutrient)

# Model results were saved as rds file to increase speed of knitting the html post
# model_lm &lt;- lm_express_mod %&gt;% 
#   mutate(model = map(data, ~lm(expression_levels ~ growth_rate, data = .x)),
#          tidy_model = map(model, tidy, conf.int = TRUE)) %&gt;% 
#   select(gene_name, tidy_model) %&gt;% 
#   unnest(tidy_model)

#write_rds(model_lm, &#39;model_tibble&#39;)
model_tib &lt;- read_rds(&quot;model_tibble&quot;)

high_express &lt;- model_tib %&gt;%
  filter(term != &quot;(Intercept)&quot;) %&gt;%
  arrange(desc(estimate)) %&gt;%
  drop_na() %&gt;% 
  filter(gene_name != &quot;&quot;) %&gt;%
  head(n=15) %&gt;% 
  mutate(gene_name_nut = glue(&quot;{gene_name}-{nutrient}&quot;)) %&gt;% 
  ggplot(aes(estimate, fct_reorder(gene_name_nut, estimate), color = estimate)) +
  geom_point() +
  geom_errorbarh(aes(xmin = conf.low, xmax = conf.high)) +
  labs(title = &quot;Genes that were &#39;turned on&#39; with increased growth rate&quot;,
       y = NULL) +
  scale_color_viridis_c() +
  theme_dark() +
  theme(legend.position = &#39;none&#39;)


low_express &lt;- model_tib %&gt;%
  filter(term != &quot;(Intercept)&quot;) %&gt;%
  drop_na() %&gt;% 
  arrange(desc(estimate)) %&gt;%
  filter(gene_name != &quot;&quot;) %&gt;%
  tail(n=15) %&gt;% 
  mutate(gene_name_nut = glue(&quot;{gene_name}-{nutrient}&quot;)) %&gt;% 
  ggplot(aes(estimate, fct_reorder(gene_name_nut, estimate), color = estimate)) +
  geom_point() +
  geom_errorbarh(aes(xmin = conf.low, xmax = conf.high)) +
  labs(title = &quot;Genes that were &#39;turned off&#39; with increased growth rate&quot;,
       y = NULL) +
  scale_color_viridis_c(begin = 1, end = 0) +
  theme_dark() +
  theme(legend.position = &#39;none&#39;)

high_express / low_express</code></pre>
<p><img src="{{< blogdown/postref >}}index_files/figure-html/unnamed-chunk-13-1.png" width="672" />
In the previous plot, we used the results from the linear model to visualize the genes with the most positive or negative slopes (estimates) respectively. We also included errorbars to show the 95% confidence intervals.</p>
<p>Let’s create one more visualization to look at the genes with the most positive change in expression in response to growth rate (under one of the nutrient conditions).</p>
<pre class="r"><code>over_expressed &lt;- c(&quot;SUL1&quot;,   &quot;UTR2&quot;,   &quot;HXT2&quot;,   &quot;CWP1&quot;,   &quot;RPL31B&quot;, &quot;RPL7A&quot;,  &quot;RPS28A&quot;, &quot;RPL24A&quot;, &quot;BUD19&quot;, &quot;CWP1&quot;,   &quot;RPS8A&quot;,  &quot;SNU13&quot;, &quot;RPP2A&quot;)

yeast_df_tidy %&gt;% 
  select(gid, gene_name, activity, molec_process, nutrient, growth_rate, expression_levels) %&gt;%
  filter(gene_name %in% over_expressed,
         gene_name != &quot;&quot;) %&gt;%
  add_count(gene_name) %&gt;%
  group_by(gene_name, gid) %&gt;%
  ggplot(aes(growth_rate, expression_levels, color = nutrient, group = nutrient)) +
  geom_line() +
  facet_wrap(~gene_name, scales = &#39;free&#39;) +
  labs(title = &quot;Expression levels vs growth rate under selective nutrient limitation&quot;,
       subtitle = &quot;Most over expressed genes&quot;,
       x = &quot;Growth rate&quot;,
       y = &quot;Expression&quot;) +                                                        
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank())</code></pre>
<p><img src="{{< blogdown/postref >}}index_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
<p>Hopefully this post has convinced you that the tidyverse can be the backbone of any omics analysis. It is essential for data cleaning, visualization, generating models and evaluating the results. That said, an important point was made in the comments of David’s post:</p>
<p>“..If you want to do statistical analysis of genomics data it is better to rely on the Bioconductor packages themselves which have been optimized for the characteristics such as small sample sizes relative to the number of features.”</p>
<p>In a follow up post, I will use an RNA-seq dataset to show how the <code>tidybulk</code> package can seamlessly bridge the tidyverse with specialized Bioconductor packages such as <code>Deseq2</code>, <code>EdgeR</code> and <code>Lima</code>. Until then, tidy on!</p>
</div>
