---
title: "Exploring the human genome with Bioconductor"
author: "Gabe Mednick"
date: '2021-10-26'
output:
  html_document:
    df_print: paged
categories: []
tags: []
subtitle: ''
summary: A primer for accessing, annotating and visualizing the human genome with Bioconductor packages
authors: []
lastmod: ''
featured: no
image:
  caption: ''
  focal_point: ''
  preview_only: no
projects: []
slug: hsapiens
---

<script src="{{< blogdown/postref >}}index_files/header-attrs/header-attrs.js"></script>
<script src="{{< blogdown/postref >}}index_files/htmlwidgets/htmlwidgets.js"></script>
<link href="{{< blogdown/postref >}}index_files/datatables-css/datatables-crosstalk.css" rel="stylesheet" />
<script src="{{< blogdown/postref >}}index_files/datatables-binding/datatables.js"></script>
<script src="{{< blogdown/postref >}}index_files/jquery/jquery-3.6.0.min.js"></script>
<link href="{{< blogdown/postref >}}index_files/dt-core/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="{{< blogdown/postref >}}index_files/dt-core/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="{{< blogdown/postref >}}index_files/dt-core/js/jquery.dataTables.min.js"></script>
<link href="{{< blogdown/postref >}}index_files/crosstalk/css/crosstalk.css" rel="stylesheet" />
<script src="{{< blogdown/postref >}}index_files/crosstalk/js/crosstalk.min.js"></script>


<div id="mission" class="section level2">
<h2>Mission</h2>
<p>Last time, we accessed data for the yeast genome and used several essential packages from the Bioconductor repository. In this post, I will load the BSgenome sequence object, genomic ranges and annotation data for the human genome and review some of the accessor functions for working with these data objects. I will use data from the <code>cpgIlands</code> package to generate visualization tracks using the <code>Gviz</code> package. To conclude, I will explore common strategies for accessing annotation data.</p>
<p>Let’s get started!</p>
<pre class="r"><code># Load packages
library(tidyverse)
theme_set(theme_light())
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(Homo.sapiens) # Use with AnnotationDbi accessors e.g., columns(Homo.sapiens)
library(Biostrings)
library(GenomicRanges)
library(BSgenome.Hsapiens.UCSC.hg19)
library(AnnotationHub)
library(plyranges)
library(GenomicFeatures)
library(ggbio)
library(Gviz)
library(glue)</code></pre>
<p>With the amazing TxDb package for the human genome (<code>TxDb.Hsapiens.UCSC.hg19.knownGene</code>), we can extract ranges for genes, transcripts, exons, promoters or other custom genomic ranges with minimal effort. We will use the <code>genes()</code> function to get the genomic ranges for all human genes.</p>
<pre class="r"><code>human_txdb &lt;- TxDb.Hsapiens.UCSC.hg19.knownGene # Genomic ragnes for genes

seqinfo(human_txdb)</code></pre>
<pre><code>## Seqinfo object with 93 sequences (1 circular) from hg19 genome:
##   seqnames       seqlengths isCircular genome
##   chr1            249250621       &lt;NA&gt;   hg19
##   chr2            243199373       &lt;NA&gt;   hg19
##   chr3            198022430       &lt;NA&gt;   hg19
##   chr4            191154276       &lt;NA&gt;   hg19
##   chr5            180915260       &lt;NA&gt;   hg19
##   ...                   ...        ...    ...
##   chrUn_gl000245      36651       &lt;NA&gt;   hg19
##   chrUn_gl000246      38154       &lt;NA&gt;   hg19
##   chrUn_gl000247      36422       &lt;NA&gt;   hg19
##   chrUn_gl000248      39786       &lt;NA&gt;   hg19
##   chrUn_gl000249      38502       &lt;NA&gt;   hg19</code></pre>
<pre class="r"><code>genes &lt;- genes(human_txdb)</code></pre>
<p>Let’s do a quick visualization of the total gene count by chromosomes. To add a layer of information, I will use the <code>glue</code> package to combine chromosome name and gene count.</p>
<pre class="r"><code>chrom_names &lt;- c(paste0(&quot;chr&quot;, 1:22), &#39;chrX&#39;, &#39;chrY&#39;)

genes %&gt;% as_tibble() %&gt;% 
  mutate(seqnames = as.character(seqnames)) %&gt;% 
  filter(seqnames %in% chrom_names) %&gt;% 
  group_by(seqnames) %&gt;% 
  count() %&gt;% 
  mutate(chrom_gene = glue(&#39;{seqnames} ({n})&#39;)) %&gt;% 
  ggplot(aes(n, fct_reorder(chrom_gene, n), fill = n)) +
  geom_col() +
  labs(title = &quot;Gene count by chromosome&quot;, 
       y = NULL,
       x = NULL) +
  theme(legend.position = &#39;none&#39;)</code></pre>
<p><img src="{{< blogdown/postref >}}index_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
<p>In the last example, I converted from GRanges to a data frame, filtered for standard chromosomes, tallied genes by chromosome and plotted the results with GGplot2. A simpler way to filter for standard chromosomes can be achieved with the function, <code>keepStandardChromosomes(x)</code> where x can be any Seqinfo object (TxDb).</p>
</div>
<div id="cpg-islands-cgis" class="section level2">
<h2>CpG islands (CGIs)</h2>
<p>CGIs are genomic regions with a higher frequency of CpG dinucleotide repeats. Many genes in vertebrate genomes have CGIs associated with their promoter regions. These CGIs are recognized as functionally important for regulating gene expression. CGIs with a lack of DNA methylation are associated with active promoter regions.</p>
<p>We will use the <code>cpgIslands</code> dataset to create an annotation data track and combine it with other tracks using the <code>Gviz</code> package. The tracks in the following plot include a representation of chromosome 7 with a red mark to specify the region of the CpG data, a ruler based on the min and max ranges of the CpG data, a track specifying the locations of CGIs and a final track for gene models in this region.</p>
<pre class="r"><code>library(Gviz)
data(cpgIslands)
cpgIslands</code></pre>
<pre><code>## GRanges object with 10 ranges and 0 metadata columns:
##        seqnames            ranges strand
##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;
##    [1]     chr7 26549019-26550183      *
##    [2]     chr7 26564119-26564500      *
##    [3]     chr7 26585667-26586158      *
##    [4]     chr7 26591772-26593309      *
##    [5]     chr7 26594192-26594570      *
##    [6]     chr7 26623835-26624150      *
##    [7]     chr7 26659284-26660352      *
##    [8]     chr7 26721294-26721717      *
##    [9]     chr7 26821518-26823297      *
##   [10]     chr7 26991322-26991841      *
##   -------
##   seqinfo: 1 sequence from hg19 genome; no seqlengths</code></pre>
<pre class="r"><code>data(geneModels)

# Data for hg19, chr7
gen &lt;- genome(cpgIslands)

# Chromosome name
chr &lt;- as.character(unique(seqnames(cpgIslands)))

# Ideogram track
itrack &lt;- IdeogramTrack(genome = gen, chromosome = chr)

# Axis ruler track
gtrack &lt;- GenomeAxisTrack()

# Add cpgIsland track
atrack &lt;- AnnotationTrack(cpgIslands, name = &quot;CpG&quot;)

# Add gene models track
grtrack &lt;- GeneRegionTrack(geneModels, genome = gen,
                           chromosome = chr, name = &quot;Gene Models&quot;)

# Combine as list plot
plotTracks(list(itrack, gtrack, atrack, grtrack))</code></pre>
<p><img src="{{< blogdown/postref >}}index_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<p>I borrowed the previous example from the <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/Gviz/inst/doc/Gviz.html">Gviz vignette</a>. Now that we are <code>Gviz</code> wizards, let’s create our own custom example using the gene info for the voltage sensing potassium (herg) channel in the human heart and surrounding regions on chromosome 2.</p>
<pre class="r"><code>herg_channel &lt;- GRanges(&quot;chr2:163227917-163695257&quot;, strand = &quot;-&quot;)

# transcriptsBy(human_txdb, herg_channel) #to get sequences
genomeAxis &lt;- GenomeAxisTrack(name=&quot;MyAxis&quot;) 


anno_track &lt;- AnnotationTrack(herg_channel, name = &quot;Herg&quot;)

# Ideogram
ideoTrack &lt;- IdeogramTrack(genome = &quot;hg19&quot;, chromosome = &quot;chr2&quot;)

hg19 &lt;- c(chr2 = &quot;hg19&quot;)
chrom = &quot;chr2&quot;

human_transcripts &lt;- transcriptsBy(human_txdb, by = &#39;gene&#39;) %&gt;% 
  as_tibble() %&gt;% 
  filter(seqnames == &quot;chr2&quot;)

geneTrack &lt;- GeneRegionTrack(human_transcripts, genome = hg19,
                           chromosome = chrom, name = &quot;Gene Models&quot;)

# Show chromosome band ID
plotTracks(list(ideoTrack, anno_track, geneTrack, genomeAxis), 
           from = 162225917, to = 164697257)</code></pre>
<p><img src="{{< blogdown/postref >}}index_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<p>There are many more track possibilities with <code>Gviz</code>. We will revisit the package in a future post and show how we can utilize many different plot types.</p>
</div>
<div id="gene-annotation" class="section level1">
<h1>Gene annotation</h1>
<p>The default gene_id column for TxDb objects uses the NCBI Entrez ID nomenclature. Before, we get into the nitty-gritty of annotating genes, it’s worth noting that we may need to change our chromosome naming conventions to work with the annotation database of interest. The package, <code>GenomeInfoDb</code> makes this easy.</p>
<pre class="r"><code>library(GenomeInfoDb)

chromNaming &lt;- genomeStyles()
names(chromNaming) # See what organism genomes are available</code></pre>
<pre><code>##  [1] &quot;Arabidopsis_thaliana&quot;     &quot;Caenorhabditis_elegans&quot;  
##  [3] &quot;Canis_familiaris&quot;         &quot;Cyanidioschyzon_merolae&quot; 
##  [5] &quot;Drosophila_melanogaster&quot;  &quot;Homo_sapiens&quot;            
##  [7] &quot;Mus_musculus&quot;             &quot;Oryza_sativa&quot;            
##  [9] &quot;Populus_trichocarpa&quot;      &quot;Rattus_norvegicus&quot;       
## [11] &quot;Saccharomyces_cerevisiae&quot; &quot;Zea_mays&quot;</code></pre>
<pre class="r"><code>chromNaming$Homo_sapiens # See conventions for homosapiens</code></pre>
<pre><code>##    circular  auto   sex NCBI  UCSC dbSNP Ensembl
## 1     FALSE  TRUE FALSE    1  chr1   ch1       1
## 2     FALSE  TRUE FALSE    2  chr2   ch2       2
## 3     FALSE  TRUE FALSE    3  chr3   ch3       3
## 4     FALSE  TRUE FALSE    4  chr4   ch4       4
## 5     FALSE  TRUE FALSE    5  chr5   ch5       5
## 6     FALSE  TRUE FALSE    6  chr6   ch6       6
## 7     FALSE  TRUE FALSE    7  chr7   ch7       7
## 8     FALSE  TRUE FALSE    8  chr8   ch8       8
## 9     FALSE  TRUE FALSE    9  chr9   ch9       9
## 10    FALSE  TRUE FALSE   10 chr10  ch10      10
## 11    FALSE  TRUE FALSE   11 chr11  ch11      11
## 12    FALSE  TRUE FALSE   12 chr12  ch12      12
## 13    FALSE  TRUE FALSE   13 chr13  ch13      13
## 14    FALSE  TRUE FALSE   14 chr14  ch14      14
## 15    FALSE  TRUE FALSE   15 chr15  ch15      15
## 16    FALSE  TRUE FALSE   16 chr16  ch16      16
## 17    FALSE  TRUE FALSE   17 chr17  ch17      17
## 18    FALSE  TRUE FALSE   18 chr18  ch18      18
## 19    FALSE  TRUE FALSE   19 chr19  ch19      19
## 20    FALSE  TRUE FALSE   20 chr20  ch20      20
## 21    FALSE  TRUE FALSE   21 chr21  ch21      21
## 22    FALSE  TRUE FALSE   22 chr22  ch22      22
## 23    FALSE FALSE  TRUE    X  chrX   chX       X
## 24    FALSE FALSE  TRUE    Y  chrY   chY       Y
## 25     TRUE FALSE FALSE   MT  chrM  chMT      MT</code></pre>
<pre class="r"><code># E.g.
seqlevelsStyle(genes) &lt;- &quot;UCSC&quot; # change chromosome naming conventions if desired
seqlevelsStyle(genes)</code></pre>
<pre><code>## [1] &quot;UCSC&quot;</code></pre>
<p>Columns, keytypes and keys are the arguments that allow <code>AnnotationDbi::select()</code> to retrieve annotation data from the OrgDb database. In the following code chunk, I use <code>columns()</code> to see the available annotation and <code>keytypes()</code> to see the types of identifiers for the <code>org.Hs.eg.db</code> database object.</p>
<pre class="r"><code>library(org.Hs.eg.db)
columns(org.Hs.eg.db) # available annotation for Homosapiens</code></pre>
<pre><code>##  [1] &quot;ACCNUM&quot;       &quot;ALIAS&quot;        &quot;ENSEMBL&quot;      &quot;ENSEMBLPROT&quot;  &quot;ENSEMBLTRANS&quot;
##  [6] &quot;ENTREZID&quot;     &quot;ENZYME&quot;       &quot;EVIDENCE&quot;     &quot;EVIDENCEALL&quot;  &quot;GENENAME&quot;    
## [11] &quot;GENETYPE&quot;     &quot;GO&quot;           &quot;GOALL&quot;        &quot;IPI&quot;          &quot;MAP&quot;         
## [16] &quot;OMIM&quot;         &quot;ONTOLOGY&quot;     &quot;ONTOLOGYALL&quot;  &quot;PATH&quot;         &quot;PFAM&quot;        
## [21] &quot;PMID&quot;         &quot;PROSITE&quot;      &quot;REFSEQ&quot;       &quot;SYMBOL&quot;       &quot;UCSCKG&quot;      
## [26] &quot;UNIPROT&quot;</code></pre>
<pre class="r"><code>keytypes(org.Hs.eg.db) # types of identifiers </code></pre>
<pre><code>##  [1] &quot;ACCNUM&quot;       &quot;ALIAS&quot;        &quot;ENSEMBL&quot;      &quot;ENSEMBLPROT&quot;  &quot;ENSEMBLTRANS&quot;
##  [6] &quot;ENTREZID&quot;     &quot;ENZYME&quot;       &quot;EVIDENCE&quot;     &quot;EVIDENCEALL&quot;  &quot;GENENAME&quot;    
## [11] &quot;GENETYPE&quot;     &quot;GO&quot;           &quot;GOALL&quot;        &quot;IPI&quot;          &quot;MAP&quot;         
## [16] &quot;OMIM&quot;         &quot;ONTOLOGY&quot;     &quot;ONTOLOGYALL&quot;  &quot;PATH&quot;         &quot;PFAM&quot;        
## [21] &quot;PMID&quot;         &quot;PROSITE&quot;      &quot;REFSEQ&quot;       &quot;SYMBOL&quot;       &quot;UCSCKG&quot;      
## [26] &quot;UNIPROT&quot;</code></pre>
<pre class="r"><code>#keys(org.Hs.eg.db) # returns keys (index) for the database contained in the OrgDb object

library(AnnotationDbi)

AnnotationDbi::select(org.Hs.eg.db, keys = &quot;A1BG&quot;, keytype = &quot;SYMBOL&quot;, 
       columns = c(&quot;SYMBOL&quot;, &quot;GENENAME&quot;, &quot;CHR&quot;, &quot;GO&quot;, &quot;PROSITE&quot;))</code></pre>
<pre><code>##    SYMBOL               GENENAME CHR         GO EVIDENCE ONTOLOGY PROSITE
## 1    A1BG alpha-1-B glycoprotein  19 GO:0002576      TAS       BP PS50835
## 2    A1BG alpha-1-B glycoprotein  19 GO:0003674       ND       MF PS50835
## 3    A1BG alpha-1-B glycoprotein  19 GO:0005576      HDA       CC PS50835
## 4    A1BG alpha-1-B glycoprotein  19 GO:0005576      IDA       CC PS50835
## 5    A1BG alpha-1-B glycoprotein  19 GO:0005576      TAS       CC PS50835
## 6    A1BG alpha-1-B glycoprotein  19 GO:0005615      HDA       CC PS50835
## 7    A1BG alpha-1-B glycoprotein  19 GO:0008150       ND       BP PS50835
## 8    A1BG alpha-1-B glycoprotein  19 GO:0031093      TAS       CC PS50835
## 9    A1BG alpha-1-B glycoprotein  19 GO:0034774      TAS       CC PS50835
## 10   A1BG alpha-1-B glycoprotein  19 GO:0043312      TAS       BP PS50835
## 11   A1BG alpha-1-B glycoprotein  19 GO:0062023      HDA       CC PS50835
## 12   A1BG alpha-1-B glycoprotein  19 GO:0070062      HDA       CC PS50835
## 13   A1BG alpha-1-B glycoprotein  19 GO:0072562      HDA       CC PS50835
## 14   A1BG alpha-1-B glycoprotein  19 GO:1904813      TAS       CC PS50835</code></pre>
<pre class="r"><code># annotating all genes
# Method 1: 
geneIDs &lt;- genes$gene_id
genes_df &lt;- AnnotationDbi::select(org.Hs.eg.db, keys = geneIDs, keytype = &quot;ENTREZID&quot;,
                                 columns = c(&quot;SYMBOL&quot;, &quot;GENENAME&quot;, &quot;ENTREZID&quot;))
genes_df %&gt;% head(30) %&gt;% DT::datatable()</code></pre>
<div id="htmlwidget-1" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-1">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30"],["1","10","100","1000","10000","100008586","100009676","10001","10002","10003","100033413","100033414","100033415","100033416","100033417","100033420","100033422","100033423","100033424","100033425","100033427","100033428","100033430","100033431","100033433","100033434","100033435","100033436","100033437","100033438"],["A1BG","NAT2","ADA","CDH2","AKT3","GAGE12F","ZBTB11-AS1","MED6","NR2E3","NAALAD2","SNORD116-1","SNORD116-2","SNORD116-3","SNORD116-4","SNORD116-5","SNORD116-8","SNORD116-10","SNORD116-11","SNORD116-12","SNORD116-13","SNORD116-15","SNORD116-16","SNORD116-18","SNORD116-20","SNORD116-22","SNORD116-23","SNORD116-24","SNORD116-25","SNORD115-2","SNORD116-26"],["alpha-1-B glycoprotein","N-acetyltransferase 2","adenosine deaminase","cadherin 2","AKT serine/threonine kinase 3","G antigen 12F","ZBTB11 antisense RNA 1","mediator complex subunit 6","nuclear receptor subfamily 2 group E member 3","N-acetylated alpha-linked acidic dipeptidase 2","small nucleolar RNA, C/D box 116-1","small nucleolar RNA, C/D box 116-2","small nucleolar RNA, C/D box 116-3","small nucleolar RNA, C/D box 116-4","small nucleolar RNA, C/D box 116-5","small nucleolar RNA, C/D box 116-8","small nucleolar RNA, C/D box 116-10","small nucleolar RNA, C/D box 116-11","small nucleolar RNA, C/D box 116-12","small nucleolar RNA, C/D box 116-13","small nucleolar RNA, C/D box 116-15","small nucleolar RNA, C/D box 116-16","small nucleolar RNA, C/D box 116-18","small nucleolar RNA, C/D box 116-20","small nucleolar RNA, C/D box 116-22","small nucleolar RNA, C/D box 116-23","small nucleolar RNA, C/D box 116-24","small nucleolar RNA, C/D box 116-25","small nucleolar RNA, C/D box 115-2","small nucleolar RNA, C/D box 116-26"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>ENTREZID<\/th>\n      <th>SYMBOL<\/th>\n      <th>GENENAME<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"order":[],"autoWidth":false,"orderClasses":false,"columnDefs":[{"orderable":false,"targets":0}]}},"evals":[],"jsHooks":[]}</script>
<pre class="r"><code># annotating all genes
# Method 2: 
genes_df2 &lt;- genes %&gt;% 
  as_tibble() %&gt;% 
  mutate(
    gene_name = AnnotationDbi::mapIds(org.Hs.eg.db,
                                      keys = genes$gene_id,
                                      keytype = &quot;ENTREZID&quot;,
                                      column = &quot;GENENAME&quot;,
                                      multiVals = &quot;first&quot;
    ),
    prot_fam = AnnotationDbi::mapIds(org.Hs.eg.db,
                                      keys = genes$gene_id,
                                      keytype = &quot;ENTREZID&quot;,
                                      column = &quot;PFAM&quot;,
                                      multiVals = &quot;first&quot;
    ))

genes_df2 %&gt;% 
  head(30) %&gt;% 
  DT::datatable()</code></pre>
<div id="htmlwidget-2" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-2">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30"],["chr19","chr8","chr20","chr18","chr1","chrX","chr3","chr14","chr15","chr11","chr15","chr15","chr15","chr15","chr15","chr15","chr15","chr15","chr15","chr15","chr15","chr15","chr15","chr15","chr15","chr15","chr15","chr15","chr15","chr15"],[58858172,18248755,43248163,25530930,243651535,49217763,101395274,71050957,72102894,89867818,25296623,25299356,25302006,25304684,25307479,25315578,25319260,25321075,25322197,25324204,25326433,25327914,25330531,25332808,25335059,25336932,25339183,25342809,25417782,25344645],[58874214,18258723,43280376,25757445,244006886,49233491,101398057,71067384,72110597,89925779,25296719,25313030,25318349,25304781,25307575,25315674,25319363,25321168,25322290,25325381,25326526,25328007,25330624,25334043,25335157,25337025,25339276,25342902,25417863,25346814],[16043,9969,32214,226516,355352,15729,2784,16428,7704,57962,97,13675,16344,98,97,97,104,94,94,1178,94,94,94,1236,99,94,94,94,82,2170],["-","+","-","-","-","+","+","-","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+"],["1","10","100","1000","10000","100008586","100009676","10001","10002","10003","100033413","100033414","100033415","100033416","100033417","100033420","100033422","100033423","100033424","100033425","100033427","100033428","100033430","100033431","100033433","100033434","100033435","100033436","100033437","100033438"],["alpha-1-B glycoprotein","N-acetyltransferase 2","adenosine deaminase","cadherin 2","AKT serine/threonine kinase 3","G antigen 12F","ZBTB11 antisense RNA 1","mediator complex subunit 6","nuclear receptor subfamily 2 group E member 3","N-acetylated alpha-linked acidic dipeptidase 2","small nucleolar RNA, C/D box 116-1","small nucleolar RNA, C/D box 116-2","small nucleolar RNA, C/D box 116-3","small nucleolar RNA, C/D box 116-4","small nucleolar RNA, C/D box 116-5","small nucleolar RNA, C/D box 116-8","small nucleolar RNA, C/D box 116-10","small nucleolar RNA, C/D box 116-11","small nucleolar RNA, C/D box 116-12","small nucleolar RNA, C/D box 116-13","small nucleolar RNA, C/D box 116-15","small nucleolar RNA, C/D box 116-16","small nucleolar RNA, C/D box 116-18","small nucleolar RNA, C/D box 116-20","small nucleolar RNA, C/D box 116-22","small nucleolar RNA, C/D box 116-23","small nucleolar RNA, C/D box 116-24","small nucleolar RNA, C/D box 116-25","small nucleolar RNA, C/D box 115-2","small nucleolar RNA, C/D box 116-26"],["PF13895","PF00797","PF00962","PF01049","PF00433","PF05831",null,"PF04934","PF00105","PF02225",null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>seqnames<\/th>\n      <th>start<\/th>\n      <th>end<\/th>\n      <th>width<\/th>\n      <th>strand<\/th>\n      <th>gene_id<\/th>\n      <th>gene_name<\/th>\n      <th>prot_fam<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[2,3,4]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
<p>I’d like to convert the protein family column from a PFAM code_id to a recognizable category (e.g., transcription factor) and plot the most common categories. I did the needful and searched Google for a bioconductor package to work with the PFAM database and found the package, <code>PFAM.db</code>. I have not figured out how to use it yet but, in a future edit, I hope to generate a plot of the most common protein families category.</p>
<pre class="r"><code>human_bsgenome &lt;- BSgenome.Hsapiens.UCSC.hg19
metadata(human_bsgenome)</code></pre>
<pre><code>## $organism
## [1] &quot;Homo sapiens&quot;
## 
## $common_name
## [1] &quot;Human&quot;
## 
## $genome
## [1] &quot;hg19&quot;
## 
## $provider
## [1] &quot;UCSC&quot;
## 
## $release_date
## [1] &quot;June 2013&quot;
## 
## $source_url
## [1] &quot;https://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/latest/&quot;</code></pre>
<pre class="r"><code>chromosomes &lt;- seqnames(human_bsgenome)

genes[genes$gene_id == &#39;10000&#39;]</code></pre>
<pre><code>## GRanges object with 1 range and 1 metadata column:
##         seqnames              ranges strand |     gene_id
##            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##   10000     chr1 243651535-244006886      - |       10000
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<pre class="r"><code>RNA_genes &lt;- genes_df2 %&gt;% # Extract genes with RNA in the gene_name
  filter(seqnames == &quot;chr8&quot; &amp; strand == &#39;+&#39;) %&gt;% 
  filter(grepl(&quot;.*[RNA].*&quot;, gene_name)) %&gt;% 
  as_granges()

chr8_rna_seqs &lt;- getSeq(human_bsgenome, names=RNA_genes)

# Another way to get sequences from a BSgenome object and a GRanges object

sample_gr &lt;- promoters(human_txdb) %&gt;% 
  head() # sample GRranges object

GenomicFeatures::getPromoterSeq(query = sample_gr, subject = human_bsgenome, upstream = 10, downstream = 10) # Designed for promoters but will work with any GRanges object</code></pre>
<pre><code>## DNAStringSet object of length 6:
##     width seq                                               names               
## [1]    20 NNNNNNNNNNNNNNNNNNNN                              uc001aaa.3
## [2]    20 NNNNNNNNNNNNNNNNNNNN                              uc010nxq.1
## [3]    20 NNNNNNNNNNNNNNNNNNNN                              uc010nxr.1
## [4]    20 CAATGCATATTTAAGGTCTA                              uc001aal.1
## [5]    20 TTACTGGTTCATTGTGAAGG                              uc001aaq.2
## [6]    20 ATGCAGTGGCTCATGCCTGT                              uc001aar.2</code></pre>
<pre class="r"><code># cytoband_chrom_plot &lt;- karyoploteR::plotKaryotype(genome=human_bsgenome)

autoplot(genes, layout = &#39;karyogram&#39;)</code></pre>
<p><img src="{{< blogdown/postref >}}index_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
<pre class="r"><code>ah &lt;- AnnotationHub()
ah &lt;- subset(ah, species == &quot;Homo sapiens&quot;)

unique(ah$dataprovider) %&gt;% 
  head(10) # Ahub data providers</code></pre>
<pre><code>##  [1] &quot;UCSC&quot;                             &quot;Ensembl&quot;                         
##  [3] &quot;RefNet&quot;                           &quot;Inparanoid8&quot;                     
##  [5] &quot;NHLBI&quot;                            &quot;NIH Pathway Interaction Database&quot;
##  [7] &quot;BroadInstitute&quot;                   &quot;Gencode&quot;                         
##  [9] &quot;MISO, VAST-TOOLS, UCSC&quot;           &quot;UWashington&quot;</code></pre>
<pre class="r"><code>unique(ah$dataprovider) %&gt;% 
  head(10) # Ahub annotation data providers</code></pre>
<pre><code>##  [1] &quot;UCSC&quot;                             &quot;Ensembl&quot;                         
##  [3] &quot;RefNet&quot;                           &quot;Inparanoid8&quot;                     
##  [5] &quot;NHLBI&quot;                            &quot;NIH Pathway Interaction Database&quot;
##  [7] &quot;BroadInstitute&quot;                   &quot;Gencode&quot;                         
##  [9] &quot;MISO, VAST-TOOLS, UCSC&quot;           &quot;UWashington&quot;</code></pre>
<pre class="r"><code>unique(ah$rdataclass) %&gt;% 
  head(10) #  Annotation data types</code></pre>
<pre><code>##  [1] &quot;GRanges&quot;          &quot;data.frame&quot;       &quot;Inparanoid8Db&quot;    &quot;TwoBitFile&quot;      
##  [5] &quot;ChainFile&quot;        &quot;SQLiteConnection&quot; &quot;biopax&quot;           &quot;BigWigFile&quot;      
##  [9] &quot;list&quot;             &quot;TxDb&quot;</code></pre>
<pre class="r"><code>mcols(ah) %&gt;% 
  head(10)</code></pre>
<pre><code>## DataFrame with 10 rows and 15 columns
##                  title dataprovider      species taxonomyid      genome
##            &lt;character&gt;  &lt;character&gt;  &lt;character&gt;  &lt;integer&gt; &lt;character&gt;
## AH5012 Chromosome Band         UCSC Homo sapiens       9606        hg19
## AH5013     STS Markers         UCSC Homo sapiens       9606        hg19
## AH5014     FISH Clones         UCSC Homo sapiens       9606        hg19
## AH5015     Recomb Rate         UCSC Homo sapiens       9606        hg19
## AH5016    ENCODE Pilot         UCSC Homo sapiens       9606        hg19
## AH5017     Map Contigs         UCSC Homo sapiens       9606        hg19
## AH5018        Assembly         UCSC Homo sapiens       9606        hg19
## AH5019 GRC Map Contigs         UCSC Homo sapiens       9606        hg19
## AH5020    Publications         UCSC Homo sapiens       9606        hg19
## AH5021   BAC End Pairs         UCSC Homo sapiens       9606        hg19
##                   description coordinate_1_based             maintainer
##                   &lt;character&gt;          &lt;integer&gt;            &lt;character&gt;
## AH5012 GRanges object from ..                  1 Marc Carlson &lt;mcarls..
## AH5013 GRanges object from ..                  1 Marc Carlson &lt;mcarls..
## AH5014 GRanges object from ..                  1 Marc Carlson &lt;mcarls..
## AH5015 GRanges object from ..                  1 Marc Carlson &lt;mcarls..
## AH5016 GRanges object from ..                  1 Marc Carlson &lt;mcarls..
## AH5017 GRanges object from ..                  1 Marc Carlson &lt;mcarls..
## AH5018 GRanges object from ..                  1 Marc Carlson &lt;mcarls..
## AH5019 GRanges object from ..                  1 Marc Carlson &lt;mcarls..
## AH5020 GRanges object from ..                  1 Marc Carlson &lt;mcarls..
## AH5021 GRanges object from ..                  1 Marc Carlson &lt;mcarls..
##        rdatadateadded          preparerclass                         tags
##           &lt;character&gt;            &lt;character&gt;                       &lt;list&gt;
## AH5012     2013-03-26 UCSCFullTrackImportP..      cytoBand,UCSC,track,...
## AH5013     2013-03-26 UCSCFullTrackImportP..        stsMap,UCSC,track,...
## AH5014     2013-03-26 UCSCFullTrackImportP..    fishClones,UCSC,track,...
## AH5015     2013-03-26 UCSCFullTrackImportP..    recombRate,UCSC,track,...
## AH5016     2013-03-26 UCSCFullTrackImportP.. encodeRegions,UCSC,track,...
## AH5017     2013-03-26 UCSCFullTrackImportP..        ctgPos,UCSC,track,...
## AH5018     2013-03-26 UCSCFullTrackImportP..          gold,UCSC,track,...
## AH5019     2013-03-26 UCSCFullTrackImportP..       ctgPos2,UCSC,track,...
## AH5020     2013-03-26 UCSCFullTrackImportP..          pubs,UCSC,track,...
## AH5021     2013-03-26 UCSCFullTrackImportP..   bacEndPairs,UCSC,track,...
##         rdataclass              rdatapath              sourceurl  sourcetype
##        &lt;character&gt;            &lt;character&gt;            &lt;character&gt; &lt;character&gt;
## AH5012     GRanges goldenpath/hg19/data.. rtracklayer://hgdown..  UCSC track
## AH5013     GRanges goldenpath/hg19/data.. rtracklayer://hgdown..  UCSC track
## AH5014     GRanges goldenpath/hg19/data.. rtracklayer://hgdown..  UCSC track
## AH5015     GRanges goldenpath/hg19/data.. rtracklayer://hgdown..  UCSC track
## AH5016     GRanges goldenpath/hg19/data.. rtracklayer://hgdown..  UCSC track
## AH5017     GRanges goldenpath/hg19/data.. rtracklayer://hgdown..  UCSC track
## AH5018     GRanges goldenpath/hg19/data.. rtracklayer://hgdown..  UCSC track
## AH5019     GRanges goldenpath/hg19/data.. rtracklayer://hgdown..  UCSC track
## AH5020     GRanges goldenpath/hg19/data.. rtracklayer://hgdown..  UCSC track
## AH5021     GRanges goldenpath/hg19/data.. rtracklayer://hgdown..  UCSC track</code></pre>
<pre class="r"><code>grs &lt;- query(ah, pattern = &quot;GRanges&quot;)
grs</code></pre>
<pre><code>## AnnotationHub with 14214 records
## # snapshotDate(): 2021-05-18
## # $dataprovider: BroadInstitute, UCSC, Ensembl, Gencode, GENCODE, EMBL-EBI, ...
## # $species: Homo sapiens
## # $rdataclass: GRanges, data.frame, DNAStringSet, GRanges
## # additional mcols(): taxonomyid, genome, description,
## #   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,
## #   rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., &#39;object[[&quot;AH5012&quot;]]&#39; 
## 
##             title                                           
##   AH5012  | Chromosome Band                                 
##   AH5013  | STS Markers                                     
##   AH5014  | FISH Clones                                     
##   AH5015  | Recomb Rate                                     
##   AH5016  | ENCODE Pilot                                    
##   ...       ...                                             
##   AH80075 | Homo_sapiens.GRCh38.100.chr.gtf                 
##   AH80076 | Homo_sapiens.GRCh38.100.chr_patch_hapl_scaff.gtf
##   AH80077 | Homo_sapiens.GRCh38.100.gtf                     
##   AH95565 | CTCF_hg19.RData                                 
##   AH95566 | CTCF_hg38.RData</code></pre>
<pre class="r"><code>#display(ah)

# Example of extracting data from the query results
res &lt;- grs[[1]] # extracting data
head(res)</code></pre>
<pre><code>## UCSC track &#39;cytoBand&#39;
## UCSCData object with 6 ranges and 1 metadata column:
##       seqnames            ranges strand |        name
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##   [1]     chr1         1-2300000      * |      p36.33
##   [2]     chr1   2300001-5400000      * |      p36.32
##   [3]     chr1   5400001-7200000      * |      p36.31
##   [4]     chr1   7200001-9200000      * |      p36.23
##   [5]     chr1  9200001-12700000      * |      p36.22
##   [6]     chr1 12700001-16200000      * |      p36.21
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<pre class="r"><code># ah2 &lt;- query(ah, c(&quot;fasta&quot;, &quot;homo sapiens&quot;, &quot;Ensembl&quot;, &quot;cdna&quot;))
# dna &lt;- ah2[[&quot;AH68262&quot;]]
# dna
# getSeq(dna)</code></pre>
</div>
